src/cluster/backend/zeromq/ZeroMQ.cc

```diff
@@ -0,0 +1,537 @@
+                }
+                else {
+                    ZEROMQ_THREAD_PRINTF("xpub: error: unexpected first char: have '0x%02x'", first);
+                    continue;
+                }
+
+                qmsgs.emplace_back(std::move(qm));
+            }
+        }
+
+        QueueForProcessing(std::move(qmsgs));
+    };
+
+    auto HandleXSubMessages = [this](const std::vector<MultipartMessage>& msgs) {
+        QueueMessages qmsgs;
+        qmsgs.reserve(msgs.size());
+
+        for ( const auto& msg : msgs ) {
+            if ( msg.size() != 4 ) {
+                ZEROMQ_THREAD_PRINTF("xsub: error: expected 4 parts, have %zu!\n", msg.size());
+                continue;
+            }
+
+            // Filter out messages that are coming from this node.
+            std::string sender(msg[1].data<const char>(), msg[1].size());
+            if ( sender == my_node_id )
+                continue;
+
+            detail::byte_buffer payload{msg[3].data<std::byte>(), msg[3].data<std::byte>() + msg[3].size()};
+            qmsgs.emplace_back(EventMessage{.topic = std::string(msg[0].data<const char>(), msg[0].size()),
+                                            .format = std::string(msg[2].data<const char>(), msg[2].size()),
+                                            .payload = std::move(payload)});
+        }
+
+        QueueForProcessing(std::move(qmsgs));
+    };
+
+    struct SocketInfo {
+        zmq::socket_ref socket;
+        std::string name;
+        std::function<void(std::vector<MultipartMessage>&)> handler;
+    };
+
+    std::vector<SocketInfo> sockets = {
+        {.socket = child_inproc, .name = "inproc", .handler = HandleInprocMessages},
+        {.socket = xpub, .name = "xpub", .handler = HandleXPubMessages},
+        {.socket = xsub, .name = "xsub", .handler = HandleXSubMessages},
+        {.socket = log_pull, .name = "log_pull", .handler = HandleLogMessages},
+    };
+
+    std::vector<zmq::pollitem_t> poll_items(sockets.size());
+
+    while ( true ) {
+        for ( size_t i = 0; i < sockets.size(); i++ )
+            poll_items[i] = {.socket = sockets[i].socket.handle(), .fd = 0, .events = ZMQ_POLLIN | ZMQ_POLLERR};
+
+        // Awkward.
+        std::vector<std::vector<MultipartMessage>> rcv_messages(sockets.size());
+        try {
+            int r = zmq::poll(poll_items, std::chrono::seconds(-1));
+            ZEROMQ_DEBUG_THREAD_PRINTF(DebugFlag::POLL, "poll: r=%d", r);
+
+            for ( size_t i = 0; i < poll_items.size(); i++ ) {
+                const auto& item = poll_items[i];
+                ZEROMQ_DEBUG_THREAD_PRINTF(DebugFlag::POLL, "poll: items[%lu]=%s %s %s\n", i, sockets[i].name.c_str(),
+                                           item.revents & ZMQ_POLLIN ? "pollin " : "",
+                                           item.revents & ZMQ_POLLERR ? "err" : "");
+
+                if ( item.revents & ZMQ_POLLERR ) {
+                    // What should we be doing? Re-open sockets? Terminate?
```

Comment:
I haven't looked at the C++ API much but I suspect somewhere we'll need to handle the usual differences of transient errors like signals vs final ones, etc. I think it's unlikely we'll nail this on the first attempt, and that's okay.

Reply:
Yeah, this is definitely a "let it happen and hopefully be loud" place.